lorien/Misc/Parser.gd

```diff
@@ -0,0 +1,197 @@
+class_name Parser
+
+# This class provides a set of inner subclasses that allow the definition
+# of a parser for a custom language. The language is defined through the means
+# of GrammarElement. Calling GrammarElement.parse("str") on a string will
+# either produce a ParsedSymbol on success, or return null if the parsing 
+# failed.
+#
+# The main Grammar elements:
+#
+# Grammar
+# 	Used to add a set of other GrammarElement objects which all are tried
+# 	to produce a valid parsing for a given string. The longest parse will be
+# 	returned as the parse of a Grammar object (Top-Down parsing). They
+#	typically form the root-element of any defined Grammar.
+#
+#	Grammar objects are initialized without arguments, and require the user
+#	to add sub-objects to their Gramar.elements array one-by-one. This is
+#	done to facilitate recursive parser elements.
+#
+#
+# GrammarSequence
+#	This class takes a list of other GrammarElement objects and tries to
+#	apply all of them in the defined sequence. A parse is considered
+#	successful only if all elements where mached successfully.
+#
+#	A GrammarSequence object takes three arguments, one optional:
+#	- The name of the symbol this GrammarSequence represents
+#	- The sequence of elements to match as an Array.
+#	- (optional) Boolean indicating a match should be "flattened" after
+#		successful parsing. Flatting means that direct symbols ancestors with
+#		the same name as the GrammarSequence will be inlined, allowing for
+#		easier definition of recursive sequence parsers
+#		(See arglist parser of the StringTemplating class for an example). 
+#
+#
+# GrammarLiteral
+# 	Class matching literal strings. Needs to be an exact match, but whitespace
+#	can be ignored through the flag GrammarLiteral.ignore_whitespace.
+#
+#
+# GrammarRegex
+# 	Same as GrammarLiteral, but uses a regex to perform the matching. If the
+#   regex contains a capture-group, the capture group will be used as the
+#	Symbol-value. Also can ignore whitespace through the 
+#	GrammarRegex.ignore_whitespace option.
+#
+#
+# ParsedSymbol
+#	If a parse was successful, this object will be return the parse-tree.
+#	The most important attributes are:
+#	- name: The symbol name that is assigned GrammarElement that was used for
+#		this node.
+#	- subsymbols: The list of symbols that are contained in this symbol. Only
+#		relevant for a GrammarSequence which will contain a list of other
+#		symbols.
+#	- value: The data that was matched by the Grammar rule.
+
+# -------------------------------------------------------------------------------------------------
+class ParsedSymbol:
+	extends Reference
+
+	var name: String
+	var last_position: int
+	var subsymbols: Array
+	var value
+	
+	func _init(_name: String, _last_position: int, _subsymbols: Array, _value).():
```

Comment:
What's `.()` doing here at the end of the line? I have never seen that in GDScript.

Reply:
It is explicitly calling the constructor of the `Reference` class. [Docs](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#class-constructor)

I am not super-sure if it is needed if the parent-constructor does receive any arguments, but I found it very important that the `Reference` class is initialized properly, because otherwise I would have introduced a big old memory leak, because the `ParsedSymbol`-instances would have not means by which they would be cleaned up, while they are used as short-lived results objects. See https://docs.godotengine.org/en/stable/tutorials/best_practices/node_alternatives.html.

Maybe we want to add this to the `Grammar*`-classes as well, but since they were meant to be used in more static user-patter, I did not bother to add it there... maybe I should... dunno. Your call ;-)

Reply:
Good to know, thanks!